### ISSUE TITLE
DatagramPacket Being Garbage Collected

### ISSUE THREAD

**Issue by hyperxpro** [netty/netty#10572]
### Expected behavior
DatagramPacket Being Garbage Collected

### Actual behavior
I store DatagramPackets into Arraylist as Backlog while DatagramChannel is established. Everything works well with a small number of PPS. As soon as I start to fire 1M PPS, GC starts collecting DatagramPackets from Arraylist and sometime even when Packets are being written into Channel `writeAndFlush`.

```
Packet got GCed -- THIS WAS DONE WHILE BACKLOG PACKETS WERE BEING WRITTEN INTO DATAGRAMCHANNEL
14 Sep 2020 19:42:30.873 [nioEventLoopGroup-2-1] ERROR i.n.u.ResourceLeakDetector - LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
#1:
	Hint: 'UpstreamHandler#0' will handle the message from this point.
	io.netty.channel.DefaultAddressedEnvelope.touch(DefaultAddressedEnvelope.java:115)
	io.netty.channel.socket.DatagramPacket.touch(DatagramPacket.java:85)
	io.netty.channel.socket.DatagramPacket.touch(DatagramPacket.java:27)
	io.netty.channel.DefaultChannelPipeline.touch(DefaultChannelPipeline.java:116)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:93)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)
#2:
	Hint: 'DefaultChannelPipeline$HeadContext#0' will handle the message from this point.
	io.netty.channel.DefaultAddressedEnvelope.touch(DefaultAddressedEnvelope.java:115)
	io.netty.channel.socket.DatagramPacket.touch(DatagramPacket.java:85)
	io.netty.channel.socket.DatagramPacket.touch(DatagramPacket.java:27)
	io.netty.channel.DefaultChannelPipeline.touch(DefaultChannelPipeline.java:116)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:93)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)
#3:
	io.netty.buffer.AdvancedLeakAwareByteBuf.internalNioBuffer(AdvancedLeakAwareByteBuf.java:736)
	io.netty.channel.socket.nio.NioDatagramChannel.doReadMessages(NioDatagramChannel.java:250)
	io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:75)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:385)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
	io.netty.buffer.AbstractByteBufAllocator.ioBuffer(AbstractByteBufAllocator.java:139)
	io.netty.channel.DefaultMaxMessagesRecvByteBufAllocator$MaxMessageHandle.allocate(DefaultMaxMessagesRecvByteBufAllocator.java:114)
	io.netty.channel.socket.nio.NioDatagramChannel.doReadMessages(NioDatagramChannel.java:246)
	io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:75)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)

```

### Steps to reproduce
Flood the server with tons of Packet.

### Minimal yet complete reproducer code (or URL to code)
https://github.com/shieldblaze/ExpressGateway/tree/l4/core/src/main/java/com/shieldblaze/expressgateway/server/udp

### Netty version
4.1.52.Final

### JVM version (e.g. `java -version`)
openjdk version "11.0.8" 2020-07-14 LTS

### OS version (e.g. `uname -a`)
Windows 10 Pro 64-Bit

**Comment by hyperxpro** [comment]
Also, why do I need to call `ByteBuf.release` when I actually want to write the packet? This is weird.

**Comment by normanmaurer** [comment]
Show me the code for `UpstreamHandler`.

**Comment by hyperxpro** [comment]
```java
final class UpstreamHandler extends ChannelInboundHandlerAdapter {

    private final List<Connection> connectionList = new CopyOnWriteArrayList<Connection>(){
        @Override
        public boolean add(Connection o) {
            super.add(o);
            Collections.sort(this);
            return true;
        }
    };

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) {
        EventLoopUtils.CHILD.next().execute(() -> {
            DatagramPacket datagramPacket = (DatagramPacket) msg;

            int index = Collections.binarySearch(connectionList, AddressUtils.address(datagramPacket.sender()), ConnectionSearchComparator.INSTANCE);

            if (index >= 0) {
                connectionList.get(index).writeDatagram(datagramPacket);
            } else {
                Connection connection = new Connection(datagramPacket.sender(), new InetSocketAddress("127.0.0.1", 9111), ctx.channel());
                connection.writeDatagram(datagramPacket);
                connectionList.add(connection);
            }
        });
    }
}

```

**Comment by hyperxpro** [comment]
```
final class Connection implements Comparable<Connection> {

    private List<DatagramPacket> datagramPacketBacklog = new ArrayList<>();
    private final byte[] clientAddressAsBytes;
    private final Channel backendChannel;
    private boolean channelActive = false;

    Connection(InetSocketAddress clientAddress, InetSocketAddress destinationAddress, Channel clientChannel) {
        clientAddressAsBytes = AddressUtils.address(clientAddress);

        Bootstrap bootstrap = BootstrapUtils.udp(EventLoopUtils.CHILD);
        bootstrap.handler(new DownstreamHandler(clientChannel, clientAddress));
        ChannelFuture channelFuture = bootstrap.connect(destinationAddress);
        backendChannel = channelFuture.channel();

        channelFuture.addListener((ChannelFutureListener) future -> {
            if (future.isSuccess()) {

                EventLoopUtils.CHILD.next().execute(() -> {
                    channelActive = true;

                    for (DatagramPacket datagramPacket : datagramPacketBacklog) {
                        if (datagramPacket == null) {
                            System.out.println("Packet got GCed");
                        } else {
                            backendChannel.writeAndFlush(datagramPacket.content());
                        }
                    }

                    datagramPacketBacklog.clear();
                });
            } else {
                channelFuture.channel().close();
                datagramPacketBacklog.clear();
            }
        });
    }

    @Override
    public int compareTo(Connection connection) {
        return SignedBytes.lexicographicalComparator().compare(clientAddressAsBytes, connection.clientAddressAsBytes);
    }

    void writeDatagram(DatagramPacket datagramPacket) {
        if (channelActive) {
            backendChannel.writeAndFlush(datagramPacket.content());
        } else {
            datagramPacketBacklog.add(datagramPacket);
        }
    }

    byte[] getClientAddressAsBytes() {
        return clientAddressAsBytes;
    }
}

```

**Comment by normanmaurer** [comment]
So I have no time to look into all of the code but from a quick look there seems to be a few problematic things:

* when channelFuture is failed you just clear the array without releasing everything that is stored in it first. This will cause a memory leak if there is anything in there at this point in time.
* Are you sure that `EventLoopUtils.CHILD.next()` will always return the same `EventLoop` ? If not then this code is not thread-safe as you modify the backing `ArrayList` from multiple threads.
* in `UpstreamHandler` you need to keep track of the connections in the ArrayList and ensure you release all the internal stored `DatagramPacket`s when `handlerRemoved` is called.

**Comment by hyperxpro** [comment]
Alright. Lemme try to debug more and see. Will keep you updated. Thanks. :)

**Comment by hyperxpro** [comment]
Your point 1: Fixed it.
Your point 2: We'll access ArrayList only once because we're just writing the pending backlog packets.
Your point 3: Yes, that can cause an issue but I've plans to implement in the future. For now, All connections are long-lived and don't disconnect. So we're good not now.

Code workflow:

Packets Arrive -> Already Connected? -> Write Packet Directly Into DatagramChannel
Packets Arrive -> Not Connected / New Connection? -> Write Packet Backlog Of Arraylist and Wait for DatagramChannel to Connect. Once connected, enable Direct write and Write all Backlog Packets from Arraylist.

```
14 Sep 2020 20:36:44.278 [nioEventLoopGroup-2-1] ERROR i.n.u.ResourceLeakDetector - LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
#1:
	Hint: 'UpstreamHandler#0' will handle the message from this point.
	io.netty.channel.DefaultAddressedEnvelope.touch(DefaultAddressedEnvelope.java:115)
	io.netty.channel.socket.DatagramPacket.touch(DatagramPacket.java:85)
	io.netty.channel.socket.DatagramPacket.touch(DatagramPacket.java:27)
	io.netty.channel.DefaultChannelPipeline.touch(DefaultChannelPipeline.java:116)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:93)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)
#2:
	Hint: 'DefaultChannelPipeline$HeadContext#0' will handle the message from this point.
	io.netty.channel.DefaultAddressedEnvelope.touch(DefaultAddressedEnvelope.java:115)
	io.netty.channel.socket.DatagramPacket.touch(DatagramPacket.java:85)
	io.netty.channel.socket.DatagramPacket.touch(DatagramPacket.java:27)
	io.netty.channel.DefaultChannelPipeline.touch(DefaultChannelPipeline.java:116)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:93)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)
#3:
	io.netty.buffer.AdvancedLeakAwareByteBuf.internalNioBuffer(AdvancedLeakAwareByteBuf.java:736)
	io.netty.channel.socket.nio.NioDatagramChannel.doReadMessages(NioDatagramChannel.java:250)
	io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:75)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:385)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
	io.netty.buffer.AbstractByteBufAllocator.ioBuffer(AbstractByteBufAllocator.java:139)
	io.netty.channel.DefaultMaxMessagesRecvByteBufAllocator$MaxMessageHandle.allocate(DefaultMaxMessagesRecvByteBufAllocator.java:114)
	io.netty.channel.socket.nio.NioDatagramChannel.doReadMessages(NioDatagramChannel.java:246)
	io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:75)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)
14 Sep 2020 20:37:14.606 [nioEventLoopGroup-3-7] WARN  i.n.u.c.DefaultPromise - An exception was thrown by com.shieldblaze.expressgateway.server.udp.Connection$$Lambda$15/1658219286.operationComplete()
java.lang.NullPointerException: null
	at com.shieldblaze.expressgateway.server.udp.Connection.lambda$null$0(Connection.java:52) ~[classes/:?]
	at java.util.ArrayList.forEach(ArrayList.java:1259) ~[?:1.8.0_265]
	at com.shieldblaze.expressgateway.server.udp.Connection.lambda$new$1(Connection.java:52) ~[classes/:?]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:604) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:104) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:300) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.connect(AbstractNioChannel.java:249) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.connect(DefaultChannelPipeline.java:1342) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:517) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.DefaultChannelPipeline.connect(DefaultChannelPipeline.java:978) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.AbstractChannel.connect(AbstractChannel.java:253) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.bootstrap.Bootstrap$3.run(Bootstrap.java:250) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-all-4.1.52.Final.jar:4.1.52.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_265]

```

In this exception, `DatagramPacket` got garbage collected while it was being written into DatagramChannel from Backlog Arraylist after future was successful. So we throw `java.lang.NullPointerException` when we tried to access `DatagramPacket#content`.

Only happening when there is a lot of GC pressure.

**Comment by hyperxpro** [comment]
@normanmaurer What do these mean?

```
14 Sep 2020 23:41:29.228 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 29 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.229 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 39 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.229 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 28 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.229 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 33 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.230 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 33 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.230 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 22 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.231 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 9 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.231 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 27 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.241 [nioEventLoopGroup-3-8] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 36 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:29.246 [nioEventLoopGroup-3-8] WARN  i.n.c.n.NioEventLoop - Selector.select() returned prematurely 512 times in a row; rebuilding Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@7107fccf.
14 Sep 2020 23:41:37.274 [nioEventLoopGroup-3-2] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 7 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@46dd4b27.
14 Sep 2020 23:41:37.274 [nioEventLoopGroup-3-2] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 4 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@46dd4b27.
14 Sep 2020 23:41:37.275 [nioEventLoopGroup-3-2] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 5 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@46dd4b27.
14 Sep 2020 23:41:37.697 [nioEventLoopGroup-3-2] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 3 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@46dd4b27.
14 Sep 2020 23:41:37.861 [nioEventLoopGroup-3-8] INFO  i.n.c.n.NioEventLoop - Migrated 2 channel(s) to the new Selector.
```

**Comment by hyperxpro** [comment]
Issue solved with `ConcurrentLinkedQueue` instead of `ArrayList`. 

Though, I'm still getting this: `[nioEventLoopGroup-3-7] DEBUG i.n.c.n.NioEventLoop - Selector.select() returned prematurely 3 times in a row for Selector io.netty.channel.nio.SelectedSelectionKeySetSelector@1a6efb4c`.

Please explain that when you're free. @normanmaurer

**Comment by normanmaurer** [comment]
there is a bug in the windows implementation of Selector and we workaround it.
